src/protocol/sender.rs

```diff
@@ -0,0 +1,622 @@
+
+    /// packet arrival rate
+    pub pkt_arr_rate: i32,
+
+    /// estimated link capacity
+    pub est_link_cap: i32,
+
+    /// Total lost packets
+    pub lost_packets: u32,
+
+    /// Total retransmitted packets
+    pub retrans_packets: u32,
+
+    /// Total received packets (packets that have been ACKed)
+    pub recvd_packets: u32,
+}
+
+impl SenderMetrics {
+    pub fn new() -> Self {
+        Self {
+            rtt: 10_000,
+            rtt_var: 0,
+            pkt_arr_rate: 0,
+            est_link_cap: 0,
+            lost_packets: 0,
+            retrans_packets: 0,
+            recvd_packets: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum SenderAlgorithmStep {
+    Step1,
+    Step2(Instant),
+    Step3(Instant),
+    Step4(Instant),
+    Step5(Instant),
+    Step6(Instant),
+}
+
+#[derive(Debug, Clone)]
+pub enum SenderAlgorithmAction {
+    Continue,
+    WaitUnitlAck,
+    WaitUntilData,
+    WaitUntil(Instant),
+    SendControl((ControlPacket, SocketAddr)),
+    SendData((DataPacket, SocketAddr)),
+    Close,
+}
+
+pub struct Sender {
+    close: bool,
+
+    current_algorithm_step: SenderAlgorithmStep,
+
+    handshake: Handshake,
+
+    /// The settings, including remote sockid and address
+    settings: ConnectionSettings,
+
+    metrics: SenderMetrics,
+
+    /// The buffer to store packets for retransmission, sorted chronologically
+    send_buffer: SendBuffer,
+
+    /// The buffer to store the next control packet
+    control_buffer: Option<ControlPacket>,
+
+    /// The buffer to store packets for transmission
+    transmit_buffer: TransmitBuffer,
+
+    // 1) Sender's Loss List: The sender's loss list is used to store the
+    //    sequence numbers of the lost packets fed back by the receiver
+    //    through NAK packets or inserted in a timeout event. The numbers
+    //    are stored in increasing order.
+    loss_list: LossList,
+
+    /// The sequence number of the largest acknowledged packet + 1
+    lr_acked_packet: SeqNumber,
+
+    /// The ack sequence number that an ack2 has been sent for
+    lr_acked_ack: i32,
+
+    /// The congestion control
+    congestion_control: SrtCongestCtrl,
+}
+
+impl Sender {
+    pub fn new(
+        settings: ConnectionSettings,
+        handshake: Handshake,
+        congestion_control: SrtCongestCtrl,
+    ) -> Self {
+        Self {
+            close: false,
+            current_algorithm_step: SenderAlgorithmStep::Step1,
+            handshake,
+            settings,
+            metrics: SenderMetrics::new(),
+            send_buffer: SendBuffer::new(&settings),
+            loss_list: LossList::new(&settings),
+            lr_acked_packet: settings.init_seq_num,
+            lr_acked_ack: -1, // TODO: why magic number?
+            congestion_control,
+            control_buffer: None,
+            transmit_buffer: TransmitBuffer::new(&settings),
+        }
+    }
+
+    pub fn next_algorithm_action<Now: Fn() -> Instant>(
+        &mut self,
```

Comment:
I would not have guessed this would take `&mut self`, maybe a different function name would be more appropriate.

Reply:
yes, I agree naming is important for facilitating API intuition, maybe "poll" would be better? But I was actually thinking of it as an iterator (it does return Option<T>...), so I chose "next"...